lib/std/sysrand.nim

```diff
@@ -0,0 +1,287 @@
+
+##
+## See also
+## ========
+## * `random module <random.html>`_
+##
+
+import std/os
+
+
+when defined(posix):
+  import std/posix
+
+  when not defined(linux):
+    const batchSize = 256
+
+    template batchImpl(result: var int, dest: var openArray[byte], getRandomImpl) =
+      let size = dest.len
+      if size == 0:
+        return
+
+      let
+        chunks = (size - 1) div batchSize
+        left = size - chunks * batchSize
+      var base = 0
+      for i in 0 ..< chunks:
+        let readBytes = getRandomImpl(addr dest[base], batchSize)
+        if readBytes < 0:
+          return readBytes
+        inc(base, batchSize)
+
+      result = getRandomImpl(addr dest[base], left)
+
+when defined(js):
+  import std/private/jsutils
+
+  when defined(nodejs):
+    {.emit: "const _nim_nodejs_crypto = require('crypto');".}
+
+    proc randomFillSync(p: Uint8Array) {.importjs: "_nim_nodejs_crypto.randomFillSync(#)".}
+
+    template urandomImpl(dest: var openArray[byte]) =
+      let size = dest.len
+      if size == 0:
+        return
+
+      var src = newUint8Array(size)
+      randomFillSync(src)
+      for i in 0 ..< size:
+        dest[i] = src[i]
+
+  else:
+    const batchSize = 256
+
+    proc getRandomValues(p: Uint8Array) {.importjs: "window.crypto.getRandomValues(#)".}
+
+    template urandomImpl(dest: var openArray[byte]) =
+      let size = dest.len
+      if size == 0:
+        return
+
+      let chunks = (size - 1) div batchSize
+      for i in 0 ..< chunks:
+        for j in 0 ..< batchSize:
+          var src = newUint8Array(batchSize)
+          getRandomValues(src)
+          dest[result + j] = src[j]
+
+        inc(result, batchSize)
+
+      let left = size - chunks * batchSize
+      var src = newUint8Array(left)
+      getRandomValues(src)
+      for i in 0 ..< left:
+        dest[result + i] = src[i]
+
+elif defined(windows):
+  type
+    PVOID = pointer
+    BCRYPT_ALG_HANDLE = PVOID
+    PUCHAR = ptr cuchar
+    NTSTATUS = clong
+    ULONG = culong
+
+  const
+    STATUS_SUCCESS = 0x00000000
+    BCRYPT_USE_SYSTEM_PREFERRED_RNG = 0x00000002
+
+  proc bCryptGenRandom(
+    hAlgorithm: BCRYPT_ALG_HANDLE,
+    pbBuffer: PUCHAR,
+    cbBuffer: ULONG,
+    dwFlags: ULONG
+  ): NTSTATUS {.stdcall, importc: "BCryptGenRandom", dynlib: "Bcrypt.dll".}
+
+
+  proc randomBytes(pbBuffer: pointer, cbBuffer: Natural): int =
+    bCryptGenRandom(nil, cast[PUCHAR](pbBuffer), ULONG(cbBuffer),
+                            BCRYPT_USE_SYSTEM_PREFERRED_RNG)
+
+  template urandomImpl(dest: var openArray[byte]) =
+    let size = dest.len
+    if size == 0:
+      return
+
+    result = randomBytes(addr dest[0], size)
+
+elif defined(linux):
+  let SYS_getrandom {.importc: "SYS_getrandom", header: "<sys/syscall.h>".}: clong
+  const syscallHeader = """#include <unistd.h>
+#include <sys/syscall.h>"""
+
+  proc syscall(
+    n: clong, buf: pointer, bufLen: cint, flags: cuint
+  ): int {.importc: "syscall", header: syscallHeader.}
```

Comment:
```suggestion
  ): clong {.importc: "syscall", header: syscallHeader.}
```
according to https://man7.org/linux/man-pages/man2/syscall.2.html

(same comment as https://github.com/nim-lang/Nim/pull/16459#pullrequestreview-575088663)